segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -652,14 +654,22 @@ private void triggerCacheTailIndex(DirectSegmentAccess segment, long tailCaching
         // Begin a loop which will end when either we've reached the target offset (defined in the RecoveryTask) or when
```

Comment:
Right now this is being triggered by `waitIfNeeded` which is checking `recoveredSegments` to see if it contains an entry or needs to have this method invoked on it. This can lead to exactly the sort of error we have observed here. If instead of just a `long`, `recoveredSegments` contained an object which contained the logic currently in `waitIfNeeded`. IE: It had a method that either invoked `toExecute` immediately or that chained it on a future. This would allow us to unify `recoveredSegments` and `recoveryTasks` and move the logic of determining if recovery has already been performed into an isolated class where it is easily checked.

Reply:
Additionally the inner class RecoveryTask should be made immutable.

Reply:
You are right and I have been able to properly address the remaining race condition that could lead to reply with stale data:
- `getBucketOffsets()` was checking the cache and then invoking `segmentTracker.waitIfNeeded()` in case nothing was in the cache. 
- By updating the cache atomically and then notifying the segment tracker, we removed the majority of problems that can be related to outstanding requests during tail-caching recovery.
- However, there was still one race condition left: Request A performs a GET and gets queued as the tail-cache needs to recovered. Then, another GET Request B gets processed. It checks the cache and it is empty because tail-caching has not yet completed. Right after that, the tail-cache recovery completes, populates the cache and notifies segment tracker about it, which also deletes the items from `recoveredSegments` and `recoveryTasks`. Then, Request B runs `getBucketOffsetFromSegment()` inside `this.segmentTracker.waitIfNeeded()`, but the `cacheUpdated` is false because the cache recovery has completed and there has been no further update to it. The result is that it checks the index, which is empty, and returns erroneously that the key does not exists.
- The solution to this is simple: any get to the tail-cache needs to be invoked as part of the callback inside `segmentTracker.waitIfNeeded()`, as it guards that tasks are properly queued or executed right away when cache is properly populated. 

Now, the test I have added passes consistently multiple times (before this change, it failed every time).

Reply:
All the members of `RecoveryTask` are already final, right? It is immutable already, unless I'm missing something.